File: python/array and linkedlist/array.py
## array
# 1. random access
def random_access(nums:list[int],index:int):
    if index<0 or index>=len(nums):
        raise Exception("Index out of bound")#抛出异常
    return nums[index]
# 2. insert
def insert(nums:list[int],index:int,value:int):
    if index<0 or index>len(nums):
        raise Exception("Index out of bound")
    nums.append(0)
    for i in range(len(nums)-1,index,-1):
        nums[i]=nums[i-1]
    nums[index]=value
# 3. delete
def delete(nums:list[int],index:int):
    if index<0 or index>=len(nums):
        raise Exception("Index out of bound")
    for i in range(index,len(nums)-1):
        nums[i]=nums[i+1]
# 4. print
def print_array(nums:list[int]):
    count  = 0
    for i in range(len(nums)):
        print(nums[i],end=' ')
    for num,index in enumerate(nums):
        count+=nums[i]
        count+=num
    print()
# 5. search
def search(nums:list[int],value:int):
    for i in range(len(nums)):
        if nums[i]==value:
            return i
    return -1
#6. extend
def extend(nums:list[int],enlarge:int)->list[int]:
    #nums.append(value)
    res=[0]*(len(nums)+enlarge)
    for i in range(len(nums)):
        res[i]=nums[i]
    return res

def main():
    arr:list[int]=[0]*5 # [0,0,0,0,0]  #分号后表示类型注解，不是必须的
    nums:list[int]=[1,2,3,4,5] 
    arr1=[0]*5 # [0,0,0,0,0]
    nums1=[1,2,3,4,5]
    print(arr1)
    print(nums1)
    print('------------------')
    ## random access
    print(random_access(nums,3))
    #print(random_access(nums,100))
    ## insert
    insert(nums,2,100)
    print(nums)
    ## delete
    delete(nums,2)
    print(nums)
    ## print
    print_array(nums)
    ## search
    print(search(nums,3))
    print(search(nums,100))
    ## extend
    print(extend(nums,5))
    print(nums)

if __name__ == "__main__":
    main()

File: cpp/Study/chapter_array_and_linkedlist/array.cpp
#include "../utils/common.hpp"
// 随机访问
int randomAccess(int *arr, int size){
    // 在区间[0, size)内随机选取一个索引
    int randomIndex = rand() % size;
    // 返回索引对应的元素
    return arr[size];
}
// 插入元素
void insert(int *nums,int size,int num,int index){
    //把索引index之后的元素都向后移动一位
    for(int i=size-1;i>index;i--){
        nums[i]=nums[i-1];
    }
    //把新元素放到索引index的位置
    nums[index]=num;
}
void Insert(int *nums,int size,int num){
    for(int i=0;i<size;i++){
        if(nums[i]>num){
            insert(nums,size,num,i);
            break;
        }
    }
}
void Inverse(int *arr){
    int len = sizeof(arr)/sizeof(arr[0]);
    for(int i=0;i<len/2;i++){
        int temp = arr[i];
        arr[i]=arr[len-i-1];
        arr[len-i-1]=temp;
    }
}
// 删除元素
void remove(int *nums,int size,int index){
    //把索引index之后的元素都向前移动一位
    for(int i=index;i<size-1;i++){
        nums[i]=nums[i+1];
    }

}
// 遍历数组
void traverse(int *nums,int size){
    for(int i=0;i<size;i++){
        cout<<nums[i]<<'\t';
    }
    cout<<endl;
}
// 查找元素
int find(int *nums,int size,int num){
    for(int i=0;i<size;i++){
        if(nums[i]==num){
            return i;
        }
    }
    return -1;
}
// 扩容数组
int *extend(int *nums,int size,int enlarge){
    int *res = new int[size+enlarge];
    for(int i=0;i<size;i++){
        res[i]=nums[i];
    }
    nums=res;
    return res;
}
int main(){
    // 创建数组
    int arr[5] = {1, 2, 3, 4, 5};
    int nums[5]={1,2,3,4,5};
    cout << arr[0]<< endl;
    cout << nums[5] << endl;
    //申请内存
    int *arr1 = new int[5];
    int *nums1 = new int[5]{1,2,3,4,5};
    cout << arr1[0]<< endl;// 输出结果为-1163005939，因为申请的内存空间没有初始化
    cout << nums1[4] << endl;
    // 释放内存
    delete[] arr1;
    delete[] nums1;
    // 随机访问
    cout<<randomAccess(arr, 5)<<endl;
    // 插入元素
    int len = sizeof(nums)/sizeof(nums[0]);    //求出数组的长度
    cout<<"-------------------"<<endl;
    traverse(nums,len);
    insert(nums,len,10000,2);
    cout<<"-------------------"<<endl;
    traverse(nums,len);
    // 删除元素
    remove(nums,len,2);
    traverse(nums,len);
    // 查找元素
    cout<<find(nums,len,3)<<endl;
    // 扩容数组
    int *nums2 = extend(nums,len,5);
    for(int i=5;i<len+5;i++){
        nums2[i]=i;
    }
    traverse(nums2,len+5);
    return 0;
}